### ISSUE TITLE
Ensure data.sql / schema.sql support is disabled automatically when using Flyway/Liquibase

### ISSUE THREAD

**Issue by philwebb** [spring-projects/spring-boot#22741]
See #20920 for background.

**Comment by ThomasVitale** [comment]
If this issue is open to community contribution, I'm available to work on it.

**Comment by wilkinsona** [comment]
@ThomasVitale Thanks very much for the offer. That'd be great, yes please. Let us know if you have any questions.

**Comment by ThomasVitale** [comment]
@wilkinsona thank you and sorry for the late answer. I looked into the code and it seems that the `DataSourceInitializer` class is the one responsible for handling the `schema.sql` and `data.sql` files. Right now, there's a condition about whether to run the scripts that is based on the value of `spring.datasource.initialization-mode` and the data source type (embedded or not). So, if I use H2 and Flyway (and I don't explicitly set the initialization mode to false), then the scripts are run. https://github.com/spring-projects/spring-boot/blob/558c2b4591b1f881b2cfb682c936eaf3aafee252/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jdbc/DataSourceInitializer.java#L123

I thought about two possible solutions that I'd like to validate with you.
A) Extending the condition in  `DataSourceInitializer` to include a check for Flyway/Liquibase.
B) Preventing the loading of `DataSourceInitializationConfiguration` if either Flyway or Liquibase is present, so that we don't register the `DataSourceInitializerPostProcessor` bean at all.

**Comment by wilkinsona** [comment]
@ThomasVitale Thanks for taking a look.

I think, ideally, `DataSourceInitializationConfiguration` would back off when Flyway or Liquibase is going to run. Unfortunately, after a brief look at the code, I'm not sure of the best way to make that happen. The auto-configuration for Flyway and Liquibase runs after the `DataSource` auto-configuration so we can't simply look for a `Flyway` or `Liquibase` bean in the context when deciding whether or not to auto-configure data source initialization.

**Comment by ThomasVitale** [comment]
Good point! Both `FlywayAutoConfiguration` and `LiquibaseAutoConfiguration` are loaded after `DataSourceAutoConfiguration`, so we can't check the beans for building the conditional. What about the properties? The autoconfiguration for Flyway and Liquibase is only loaded if `spring.flyway.enabled=true` and `spring.liquibase.enabled=true`, respectively. What if we checked for the presence of Flyway/Liquibase in the classpath and whether the corresponding property is true? Though, it would probably not be the cleanest solution.

**Comment by wilkinsona** [comment]
> What if we checked for the presence of Flyway/Liquibase in the classpath and whether the corresponding property is true? Though, it would probably not be the cleanest solution.

I agree that may not be the cleanest. I think we'd be in danger of duplicating the Flyway and Liquibase auto-configuration conditions in the DataSource initialization auto-configuration.

The core team discussed this one yesterday and decided that it's something that should be tackled as part of a broader revamp of DataSource initialization that we'd like to do in 3.0. Thanks anyway for taking a look.

**Comment by wilkinsona** [comment]
Assuming I have understood your use case correctly, it is already supported by Flyway: https://flywaydb.org/documentation/concepts/migrations.html#schema-creation. I think that within the database migration tool itself is the right place for this sort of functionality as it's applicable to anyone using Flyway not just those using Flyway with Spring Boot.

**Comment by lbruun** [comment]
@wilkinsona. Yes, I think you've understood correctly. Unfortunately I'm using using Liquibase, not Flyway, and wasn't aware of this feature in Flyway. There's no equivalent in Liquibase.  Flyway has the "advantage" that it doesn't attempt to be database engine agnostic. Liquibase does indeed attempt to be database engine agnostic and such a feature would open a myriad of questions for them, AFAIU. I can see it has been discussed on several occasions in Liquibase Forum ([Ref1](https://forum.liquibase.org/t/advice-on-how-to-create-a-schema-for-defaultschemaname/470), [Ref2](https://forum.liquibase.org/t/how-to-create-a-schema-for-defaultschemaname/4103)). It seems like Nathan, the original developer of Liquibase, would like to keep it "clean" and is therefore hesitant.

I'll ponder a bit if I agree with you or not that it doesn't belong in Spring Boot. :-) I'll go back to thinking.

**Comment by ThomasVitale** [comment]
In Flyway, you can also define callbacks (defined in SQL or Java) that let you hook into the Flyway lifecyle and execute specific procedures (for example, before the baseline migration is run or before automatically creating non-existent schemas). https://flywaydb.org/documentation/concepts/callbacks

**Comment by itstannus** [comment]
> @wilkinsona. Yes, I think you've understood correctly. Unfortunately I'm using using Liquibase, not Flyway, and wasn't aware of this feature in Flyway. There's no equivalent in Liquibase. Flyway has the "advantage" that it doesn't attempt to be database engine agnostic. Liquibase does indeed attempt to be database engine agnostic and such a feature would open a myriad of questions for them, AFAIU. I can see it has been discussed on several occasions in Liquibase Forum ([Ref1](https://forum.liquibase.org/t/advice-on-how-to-create-a-schema-for-defaultschemaname/470), [Ref2](https://forum.liquibase.org/t/how-to-create-a-schema-for-defaultschemaname/4103)). It seems like Nathan, the original developer of Liquibase, would like to keep it "clean" and is therefore hesitant.
> 
> I'll ponder a bit if I agree with you or not that it doesn't belong in Spring Boot. :-) I'll go back to thinking.

@Ibrunn I am also working with Liquibase , did you find  the solution for this problem?

**Comment by lbruun** [comment]
@itstannus : Yeah, my solution is the [Pre-Liquibase Spring Boot module](https://github.com/lbruun/Pre-Liquibase). But I think you already found it.  :-)

In a few words it is a Spring Boot autoconfig which essentially fires SQL scripts and guaranteed to execute _before_ Spring Boot invokes Liquibase itself.
